flax/nnx/nnx/bridge/variables.py

```diff
@@ -0,0 +1,131 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, TypeVar
+
+import jax
+from flax import struct
+from flax.core import meta
+from flax.nnx.nnx import variables as variableslib
+import typing as tp
+
+
+A = TypeVar('A')
+B = TypeVar('B')
+
+
+#######################################################
+### Variable type <-> Linen collection name mapping ###
+#######################################################
+# Assumption: the mapping is 1-1 and unique.
+
+VariableTypeCache: dict[str, tp.Type[variableslib.Variable[tp.Any]]] = {}
+
+
+def variable_type(name: str) -> tp.Type[variableslib.Variable[tp.Any]]:
+  """Given a Linen-style collection name, get or create its corresponding NNX Variable type."""
+  if name not in VariableTypeCache:
+    VariableTypeCache[name] = type(name, (variableslib.Variable,), {})
+  return VariableTypeCache[name]
+
+
+def variable_type_name(typ: tp.Type[variableslib.Variable[tp.Any]]) -> str:
+  """Given an NNX Variable type, get or create its Linen-style collection name.
+
+  Should output the exact inversed result of `variable_type()`."""
+  for name, t in VariableTypeCache.items():
+    if typ == t:
+      return name
+  name = typ.__name__
+  if name in VariableTypeCache:
+    raise ValueError(
+      'Name {name} is already registered in the registry as {VariableTypeCache[name]}. '
+      'It cannot be linked with this type {typ}.'
+    )
+  register_variable_name_type_pair(name, typ)
+  return name
+
+
+def register_variable_name_type_pair(name, typ):
+  """Register a pair of variable type name (like Linen collections) and its NNX type."""
+  VariableTypeCache[name] = typ
+
+
+# add known variable type names
+register_variable_name_type_pair('params', variableslib.Param)
+register_variable_name_type_pair('batch_stats', variableslib.BatchStat)
+register_variable_name_type_pair('cache', variableslib.Cache)
+register_variable_name_type_pair('intermediates', variableslib.Intermediate)
+
+
+def sort_variable_types(types: list[type]):
+  def _variable_parents_count(t: type):
+    return sum(1 for p in t.mro() if issubclass(p, variableslib.Variable))
+  parent_count = {t: _variable_parents_count(t) for t in types}
+  return sorted(types, key=lambda t: -parent_count[t])
+
+
+#############################################
+### NNX Variable <-> Linen metadata boxes ###
+#############################################
+
+
+class NNXMeta(struct.PyTreeNode, meta.AxisMetadata[A]):
+  """Default Flax metadata class for `nnx.VariableState`.
+  """
+
+  var_type: type[variableslib.Variable[tp.Any]] = struct.field(pytree_node=False)
+  value: Any = struct.field(pytree_node=True)
+  metadata: dict[str, tp.Any] = struct.field(pytree_node=False)
+
+  def unbox(self) -> A:
+    return self.value
+
+  def replace_boxed(self, val: B) -> 'NNXMeta[B]':
```

Comment:
String types not needed in python >= 3.10
```suggestion
  def replace_boxed(self, val: B) -> NNXMeta[B]:
```

Reply:
I got error trying to do this, saying `NNXMeta is not found`. I'll leave it as is for now.